rfcs/20191127-pip-structure.md

```diff
@@ -0,0 +1,212 @@
+`from tensorflow.foo` imports but lives under tensorflow\_core package. Simplifying
+package structure would improve productivity for TensorFlow users.
+
+## Design Proposal
+
+The best way I can think of to fix the autocomplete issues is to make our package structure as clean as possible. In this case, autocomplete will work out of the box.
+
+### Short term: Remove virtual pip package
+
+Primary purpose of keeping the virtual pip package is to workaround circular
+estimator imports. Alternatively, we can resolve this issue by lazy loading
+estimator.
+
+Lazy loading by itself would mean that we no longer have autocomplete for estimator. As a workaround, we can import estimator in the stub *.pyi* file.
+
+Estimator import in root *\_\_init\_\_.py* file:
+```python
+from tensorflow.python.util.lazy_loader import LazyLoader as _LazyLoader
+estimator = _LazyLoader(
+    "estimator", globals(),
+    "tensorflow_estimator.python.estimator.api._v2.estimator")
+setattr(_current_module, "estimator", estimator)
+```
+
+Estimator import in root *\_\_init\_\_.pyi* file:
+```python
+from tensorflow_estimator.python.estimator.api._v2 import estimator as estimator
+```
+In Python 3, we don't even need the *.pyi* file and instead can import estimator
+without lazy loading if `typing.TYPE_CHECKING` is `True`.
+
+After building a pip package with this change all of the following work in PyCharms (both released and EAP) and VS Code:
+
+* jump-to-definition
+* quick documentation
+* autocomplete for `compat.v1.keras`, `compat.v2.keras`
+* autocomplete for keras when using from tensorflow import keras
+* ...basically any import I tested works with autocompletion
+
+To support the TensorFlow Metapackage plans we could add a new pip package that specifies dependencies on tensorflow, tensorflow\_estimator, tensorboard, etc.. Its sole purpose would be to get all dependencies installed.
+
+![alt_text](https://github.com/annarev/community/blob/pip_structure_rfc/rfcs/20191127-pip-structure/new_modular_structure.png "New proposed modular TensorFlow structure.")
+
+### Long term: Import from external package directly
+Short term would fix IDE issues, but the package structure is still not as clean as it could be. We resolve cycles with lazy loading but it would be even better not to have this circular structure at all.
+
+Therefore, I propose that we don’t import external packages into TensorFlow 3.0. Users who want to use estimator, tensorboard summaries or keras could import them separately:
+
+Current code that looks like:
+```python
+import tensorflow as tf
+
+tf.estimator
+tf.keras
+tf.summary
+```
+
+Would be changed to:
+```python
+import tensorflow as tf
+import tensorflow_estimator as estimator
+import tensorflow_keras as keras
+from tenosorboard import summaries
+```
+
+Rationale for this change:
+
+* One way dependencies (estimator depends on tensorflow and not vise-versa).
+* Minimal overhead for users. Adding an extra import is easy.
```

Comment:
Myself I vote against (if I can get a vote :-) Adding an extra import is easy, but adding 3 additional imports is less so, and add an import each time you use more and more functionality is bothersome. Especially `keras` and `tensorboard` seem basic part of TF 2 (at least in my view), and importing them from another package seems wrong.

On the next line you talk about keeping aliases like `tf.summary` -- if you plan to support that, than I would be grateful. But maybe the plan is to deprecate and move users to individual imports?

Reply:
Primary reason for this is to avoid circular structure. keras, estimator and tensorboard currently all depend on tensorflow pip package (they would use import tensorflow.*). If we make tf.estimator, tf.keras and tf.summary available (which is the case now), then we have a reverse dependency as well. This is a cycle.

That being said, I am definitely interested in hearing opinions about having these separate imports! If tf.keras and tf.estimator are very useful, we could consider ways to deal with circular imports.

There are two ways I can think of to deal with circular imports now:
1. Lazy loading.
2. There is another option in Alternatives Considered right below. Basically, instead of having these separate imports, we can rename tensorflow package to tensorflow_core. estimator, keras and tensorboard would then use "from tensorflow_core.". Then we have another package on top called tensorflow that will combine tensorflow_core, estimator, keras and tensorboard into one package.

Both of these options seem less clean to me than just having the extra imports, but I am open to other options based on feedback.

Reply:
The current plan is to make it possible to import from stand alone packages (for e.g. tensorflow_estimator and tensorflow_keras) but at the same time keep `tf.estimator` and `tf.keras` available through lazy loading.